### DISCUSSION TITLE
Microservice to manage AWS roles, built with Fast API

### DISCUSSION THREAD

**Discussion by michaeljcollinsuk** [ministryofjustice/analytical-platform#2588] [Firebreak Winter 23]
### Idea Description

Microservice to provide an API with Fast API to manage AWS roles

### Why Should We Do This

Currently AWS permissions (such as s3 bucket access) are tightly coupled to the Analytical Platform Control Panel. This is not ideal for many reasons:
- the structure of the AWS roles are hardcoded within the AP Control Panel codebase
- updating these roles may require code changes in Control Panel and a new release
- requires knowledge of the AP Control Panel
- poor visibility of changes to AWS roles in the wider team
- the wider functionality of the AP Control Panel may not always be maintained

From a team/organisation perspective, by decoupling management of AWS roles to a new microservice, in future we would be able to manage users AWS permissions from other products/services rather than just the AP Control Panel. Initially as a POC this could mean updating the AP control panel to communicate with the AWS microservice in order to update users S3 access. In the future it could be integrating with another service (e.g. granting athena access to a registered data product).

From a personal development perspective, I am interested in learning Fast API as I have not worked with the technology before, and think it could be useful for future projects.

### Definition of Done
- [ ] Deployed microservice
- [ ] Integration with another service (e.g. control panel) to perform AWS-role management via API endpoints and task/queue approach(optional)

### What skills do we need?
Developers and engineers to help with application level code, build and deployment and (optionally) CI workflows

**Comment by ymao2** [comment]
I join 😺

**Comment by michaeljcollinsuk** [comment]
Slack channel: https://moj.enterprise.slack.com/archives/C0694SMLVUK
